src/core/internal/array/equality.d

```diff
@@ -10,20 +10,56 @@
 
 module core.internal.array.equality;
 
+ // `lhs == rhs` lowers to `__equals(lhs, rhs)` for dynamic arrays
+ // The scalar-only overload takes advantage of known properties of scalars to
+ // consolidate IFTI instances and remove the need for helper functions.
+bool __equals(T1, T2)(scope const T1[] lhs, scope const T2[] rhs)
+@nogc nothrow pure @trusted
+if (__traits(isScalar, T1) && __traits(isScalar, T2))
+{
+    if (lhs.length != rhs.length)
+        return false;
+    static if (T1.sizeof == T2.sizeof
+        // Signedness needs to match for types that promote to int.
+        && (T1.sizeof >= 4 || __traits(isUnsigned, T1) == __traits(isUnsigned, T2))
+        && !__traits(isFloating, T1) && !__traits(isFloating, T2))
+    {
+        if (!__ctfe)
+        {
+            // This would improperly allow equality of integers and pointers
+            // but the CTFE branch will stop this function from compiling then.
+            import core.stdc.string : memcmp;
+            return lhs.length == 0 || 0 == memcmp(lhs.ptr, rhs.ptr, lhs.length * T1.sizeof);
```

Comment:
not sure you need the length == 0 case.

Reply:
it's an optimisation, for an uncommon case, therefore a slowdown for the common case.

Reply:
It's there because I am under the impression that `memcmp` has undefined behavior for null pointers. If not it can be removed.

Reply:
@n8sh you are correct.
memcpy on null is undefined